File: asherah-ruby/lib/asherah.rb
# frozen_string_literal: true

require "json"

require_relative "asherah/version"
require_relative "asherah/error"
require_relative "asherah/config"
require_relative "asherah/native"
require_relative "asherah/session_factory"
require_relative "asherah/session"
require_relative "asherah/hooks"

module Asherah
  @mutex = Mutex.new
  @factory = nil
  @sessions = {}
  @initialized = false
  @session_cache_enabled = true
  @verbose = false

  class << self
    # Configure Asherah using a block with snake_case accessors.
    # Compatible with the canonical godaddy/asherah-ruby gem API.
    #
    #   Asherah.configure do |config|
    #     config.service_name = "MyService"
    #     config.product_id = "MyProduct"
    #     config.kms = "static"
    #     config.metastore = "memory"
    #   end
    def configure
      @mutex.synchronize do
        raise Error::AlreadyInitialized if @initialized

        config = Config.new
        yield config
        config.validate!

        json = config.to_json
        pointer = Native.asherah_factory_new_with_config(json)
        @factory = SessionFactory.new(pointer)
        @sessions = {}
        @initialized = true
        @session_cache_enabled = config.enable_session_caching != false
        @verbose = config.verbose == true
      end
    end

    # Initialize Asherah with a PascalCase config hash.
    # Also accepts snake_case string/symbol keys (auto-normalized).
    def setup(config)
      normalized = normalize_config(config)
      json = JSON.generate(normalized)

      pointer = Native.asherah_factory_new_with_config(json)
      factory = SessionFactory.new(pointer)

      @mutex.synchronize do
        raise Error::AlreadyInitialized if @initialized

        @factory = factory
        @sessions = {}
        @initialized = true
        @session_cache_enabled = truthy(normalized["EnableSessionCaching"], default: true)
        @verbose = truthy(normalized["Verbose"], default: false)
      end

      nil
    rescue StandardError
      factory&.close if defined?(factory) && factory
      raise
    end

    def setup_async(config, &block)
      Thread.new do
        result = setup(config)
        block&.call(result)
        result
      end
    end

    def shutdown
      factory = nil
      sessions = nil
      @mutex.synchronize do
        raise Error::NotInitialized unless @initialized

        factory = @factory
        sessions = @sessions.values
        @factory = nil
        @sessions = {}
        @initialized = false
      end

      Array(sessions).each do |session|
        begin
          session.close unless session.closed?
        rescue StandardError => e
          warn "asherah: error closing session during shutdown: #{e.message}"
        end
      end
      factory&.close unless factory&.closed?
      nil
    end

    def shutdown_async(&block)
      Thread.new do
        result = shutdown
        block&.call(result)
        result
      end
    end

    def get_setup_status
      @mutex.synchronize { @initialized }
    end

    def setenv(env = {})
      data = case env
             when String
               JSON.parse(env)
             else
               env
             end
      unless data.respond_to?(:each_pair)
        raise ArgumentError, "environment payload must be a Hash or JSON object"
      end
      data.each_pair do |k, v|
        if v.nil?
          ENV.delete(String(k))
        else
          ENV[String(k)] = v.to_s
        end
      end
      nil
    end
    alias_method :set_env, :setenv

    def encrypt(partition_id, payload)
      raise ArgumentError, "payload cannot be nil" if payload.nil?
      session = resolve_session(partition_id)
      session.encrypt_bytes(payload)
    end

    def encrypt_string(partition_id, text)
      raise ArgumentError, "text cannot be nil" if text.nil?
      encrypt(partition_id, text)
    end

    def decrypt(partition_id, data_row_record)
      raise ArgumentError, "data_row_record cannot be nil" if data_row_record.nil?
      session = resolve_session(partition_id)
      session.decrypt_bytes(data_row_record).force_encoding(Encoding::UTF_8)
    end

    def decrypt_string(partition_id, data_row_record)
      raise ArgumentError, "data_row_record cannot be nil" if data_row_record.nil?
      decrypt(partition_id, data_row_record).force_encoding(Encoding::UTF_8)
    end

    def encrypt_async(partition_id, payload, &block)
      Thread.new do
        result = encrypt(partition_id, payload)
        block&.call(result)
        result
      end
    end

    def decrypt_async(partition_id, data_row_record, &block)
      Thread.new do
        result = decrypt(partition_id, data_row_record)
        block&.call(result)
        result
      end
    end

    # Install a log hook. Yields a +Hash+ +{level:, target:, message:}+ for
    # every log record emitted by the underlying Rust crates. The block may
    # fire from any thread; implementations must be thread-safe and
    # non-blocking. Pass +nil+ to clear (equivalent to {clear_log_hook}).
    #
    # Replaces any previously installed log hook. Exceptions raised from the
    # callback are caught and silently swallowed.
    def set_log_hook(callback = nil, &block)
      Hooks.set_log_hook(callback, &block)
    end

    # Remove the active log hook, if any. Idempotent.
    def clear_log_hook
      Hooks.clear_log_hook
    end

    # Install a metrics hook. Yields a +Hash+ +{type:, duration_ns:, name:}+
    # for every metrics event. Timing events ({:encrypt, :decrypt, :store,
    # :load}) carry a positive +duration_ns+ and a +nil+ +name+; cache events
    # ({:cache_hit, :cache_miss, :cache_stale}) carry +duration_ns+ == 0 and
    # the cache identifier in +name+.
    #
    # Installing a hook implicitly enables the global metrics gate; clearing
    # it disables the gate. Replaces any previously installed metrics hook.
    # Pass +nil+ to clear (equivalent to {clear_metrics_hook}).
    def set_metrics_hook(callback = nil, &block)
      Hooks.set_metrics_hook(callback, &block)
    end

    # Remove the active metrics hook and disable metrics. Idempotent.
    def clear_metrics_hook
      Hooks.clear_metrics_hook
    end

    private

    REQUIRED_KEYS = %w[ServiceName ProductID Metastore].freeze

    def normalize_config(config)
      unless config.respond_to?(:each_pair)
        raise ArgumentError, "config must be a Hash-like object"
      end
      normalized = {}
      config.each_pair do |key, value|
        normalized[String(key)] = value
      end
      REQUIRED_KEYS.each do |key|
        raise ArgumentError, "#{key} is required" if normalized[key].nil? || normalized[key].to_s.strip.empty?
      end
      normalized
    end

    def truthy(value, default: false)
      return default if value.nil?

      case value
      when true, "1", "true", "TRUE", "yes", "on" then true
      when false, "0", "false", "FALSE", "no", "off" then false
      else
        default
      end
    end

    def resolve_session(partition_id)
      raise ArgumentError, "partition_id cannot be empty" if String(partition_id).empty?

      # Brief mutex hold for hash lookup only — FFI call happens outside
      @mutex.synchronize do
        raise Error::NotInitialized unless @initialized
        if @session_cache_enabled
          @sessions[partition_id] ||= @factory.get_session(partition_id)
        else
          @factory.get_session(partition_id)
        end
      end
    end
  end
end

File: asherah-php/src/Asherah.php
<?php

declare(strict_types=1);

namespace GoDaddy\Asherah;

/**
 * @phpstan-type AsherahConfigArray array{
 *   ServiceName?: string,
 *   ProductID?: string,
 *   Metastore?: string,
 *   KMS?: string,
 *   ConnectionString?: string,
 *   SQLMetastoreDBType?: string,
 *   ReplicaReadConsistency?: string,
 *   DynamoDBTableName?: string,
 *   DynamoDBRegion?: string,
 *   DynamoDBSigningRegion?: string,
 *   DynamoDBEndpoint?: string,
 *   EnableRegionSuffix?: bool,
 *   RegionMap?: array<string, string>,
 *   PreferredRegion?: string,
 *   KmsKeyId?: string,
 *   StaticMasterKeyHex?: string,
 *   AwsProfileName?: string,
 *   EnableSessionCaching?: bool,
 *   SessionCacheMaxSize?: int,
 *   SessionCacheDuration?: int,
 *   ExpireAfter?: int,
 *   CheckInterval?: int,
 *   ...<string, mixed>
 * }
 */
final class Asherah
{
    private static ?SessionFactory $factory = null;
    /** @var array<string, Session> */
    private static array $sessions = [];
    private static int $sessionCacheMaxSize = 1000;
    private static bool $sessionCacheEnabled = true;

    /**
     * @param AsherahConfigArray|AsherahConfig $config
     */
    public static function setup(array|AsherahConfig $config): void
    {
        if (self::$factory !== null) {
            throw new LifecycleException('already initialized');
        }

        $config = self::normalizeConfig($config);
        foreach (['ServiceName', 'ProductID', 'Metastore', 'KMS'] as $key) {
            if (!isset($config[$key]) || trim((string) $config[$key]) === '') {
                throw new ConfigurationException("$key is required");
            }
        }

        self::$sessionCacheEnabled = self::sessionCacheEnabled($config);
        self::$sessionCacheMaxSize = self::sessionCacheMaxSize($config);
        self::validateIntegerOption($config, 'SessionCacheDuration', 0);
        self::validateIntegerOption($config, 'ExpireAfter', 1);
        self::validateIntegerOption($config, 'CheckInterval', 1);
        self::$factory = SessionFactory::fromConfig($config);
        self::$sessions = [];
    }

    public static function shutdown(): void
    {
        foreach (self::$sessions as $session) {
            $session->close();
        }
        self::$sessions = [];
        self::$factory?->close();
        self::$factory = null;
    }

    public static function encrypt(string $partitionId, string $payload): string
    {
        if (!self::$sessionCacheEnabled) {
            $session = self::newSession($partitionId);
            try {
                return $session->encryptBytes($payload);
            } finally {
                $session->close();
            }
        }

        return self::cachedSession($partitionId)->encryptBytes($payload);
    }

    public static function encryptBytes(string $partitionId, string $payload): string
    {
        return self::encrypt($partitionId, $payload);
    }

    public static function decrypt(string $partitionId, string $dataRowRecord): string
    {
        if (!self::$sessionCacheEnabled) {
            $session = self::newSession($partitionId);
            try {
                return $session->decryptBytes($dataRowRecord);
            } finally {
                $session->close();
            }
        }

        return self::cachedSession($partitionId)->decryptBytes($dataRowRecord);
    }

    public static function decryptBytes(string $partitionId, string $dataRowRecord): string
    {
        return self::decrypt($partitionId, $dataRowRecord);
    }

    public static function encryptString(string $partitionId, string $payload): string
    {
        return self::encrypt($partitionId, $payload);
    }

    public static function decryptString(string $partitionId, string $dataRowRecord): string
    {
        return self::decrypt($partitionId, $dataRowRecord);
    }

    private static function cachedSession(string $partitionId): Session
    {
        if (isset(self::$sessions[$partitionId])) {
            $session = self::$sessions[$partitionId];
            unset(self::$sessions[$partitionId]);
            self::$sessions[$partitionId] = $session;
            return $session;
        }

        $session = self::newSession($partitionId);
        self::$sessions[$partitionId] = $session;
        if (count(self::$sessions) > self::$sessionCacheMaxSize) {
            $evicted = array_key_first(self::$sessions);
            if ($evicted !== null) {
                $old = self::$sessions[$evicted];
                unset(self::$sessions[$evicted]);
                $old->close();
            }
        }

        return $session;
    }

    private static function newSession(string $partitionId): Session
    {
        if ($partitionId === '') {
            throw new \InvalidArgumentException('partition_id cannot be empty');
        }
        if (self::$factory === null) {
            throw new LifecycleException('not initialized');
        }

        return self::$factory->getSession($partitionId);
    }

    /**
     * @param array<string, mixed>|AsherahConfig $config
     * @return array<string, mixed>
     */
    private static function normalizeConfig(array|AsherahConfig $config): array
    {
        return $config instanceof AsherahConfig ? $config->toArray() : $config;
    }

    /**
     * @param array<string, mixed> $config
     */
    private static function sessionCacheEnabled(array $config): bool
    {
        if (!array_key_exists('EnableSessionCaching', $config)) {
            return true;
        }
        if (!is_bool($config['EnableSessionCaching'])) {
            throw new ConfigurationException('EnableSessionCaching must be boolean');
        }

        return $config['EnableSessionCaching'];
    }

    /**
     * @param array<string, mixed> $config
     */
    private static function sessionCacheMaxSize(array $config): int
    {
        if (!array_key_exists('SessionCacheMaxSize', $config)) {
            return 1000;
        }
        if (!is_int($config['SessionCacheMaxSize']) || $config['SessionCacheMaxSize'] < 1) {
            throw new ConfigurationException('SessionCacheMaxSize must be an integer >= 1');
        }

        return $config['SessionCacheMaxSize'];
    }

    /**
     * @param array<string, mixed> $config
     */
    private static function validateIntegerOption(array $config, string $key, int $minimum): void
    {
        if (!array_key_exists($key, $config)) {
            return;
        }
        if (!is_int($config[$key]) || $config[$key] < $minimum) {
            throw new ConfigurationException("{$key} must be an integer >= {$minimum}");
        }
    }
}
